puma/hlplots/tagger.py

```diff
@@ -0,0 +1,214 @@
+            passing a file path to a h5 file with a structured numpy array,
+            by default "data_frame"
+        key : str, optional
+            Key within h5 file, needs to be provided when using the `source_type`
+            `data_frame_path` or `numpy_structured`, by default None
+        Raises
+        ------
+        ValueError
+            if source_type is wrongly specified
+        """
+        # list tagger variables
+        tagger_vars = [f"{self.model_name}_{flv}" for flv in self.flvs]
+        # TODO: change to cae syntax in python 3.10
+        if source_type == "data_frame":
+            logger.debug("Retrieving tagger `%s` from data frame.", self.model_name)
+            self.scores = source[tagger_vars].values
+            return
+        if key is None:
+            raise ValueError(
+                "When using a `source_type` other than `data_frame`, you need to "
+                "specify the `key`."
+            )
+        if source_type == "data_frame_path":
+            logger.debug(
+                "Retrieving tagger %s in data frame from file %s.",
+                self.model_name,
+                source,
+            )
+            df_in = pd.read_hdf(source, key=key)
+            self.scores = df_in[tagger_vars].values
+
+        elif source_type == "numpy_structured":
+            logger.debug(
+                "Retrieving tagger %s from structured numpy file %s.",
+                self.model_name,
+                source,
+            )
+            with h5py.File(source, "r") as f_h5:
+                self.scores = structured_to_unstructured(
+                    f_h5[key].fields(tagger_vars)[:]
+                )
+        else:
+            raise ValueError(f"{source_type} is not a valid value for `source_type`.")
+
+    def _init_from_template(self, template):
+        if template is not None:
+            for key, val in template.items():
+                if hasattr(self, key):
+                    setattr(self, key, val)
+                else:
+                    raise KeyError(f"`{key}` is not an attribute of the Tagger class.")
+        else:
+            logger.debug(
+                "Template initialised with template being `None` - not doing anything."
+            )
+
+    @property
+    def n_jets_light(self):
+        """Retrieve number of light jets.
+
+        Returns
+        -------
+        int
+            number of light jets
+        """
+        return int(np.sum(self.is_light))
+
+    @property
+    def n_jets_c(self):
+        """Retrieve number of c jets.
+
+        Returns
+        -------
+        int
+            number of c jets
+        """
+        return int(np.sum(self.is_c))
+
+    @property
+    def n_jets_b(self):
+        """Retrieve number of b jets.
+
+        Returns
+        -------
+        int
+            number of b jets
+        """
+        return int(np.sum(self.is_b))
+
+
+class Tagger(TaggerBase):
```

Comment:
Why have base and derived classes here?

Reply:
I can merge them back, just thought it might be more readable like that, also the linter was complaining about too many instances

Reply:
I don't have a strong preference either way (maybe a mild preference for merging them as I think having a few more instances is cleaner than inheritance). mainly just wondered if there was a more architectural reason.

Reply:
I merged them now